runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java

```diff
@@ -198,28 +330,62 @@ String updateMessage(final String message)
 		return Strings.join(messageWords, " ");
 	}
 
-	/**
-	 * Remove tags, except for &lt;lt&gt; and &lt;gt&gt;
-	 *
-	 * @return
-	 */
-	private static String removeTags(String str)
+	private void timeoutEmojiMessages(boolean force)
 	{
-		StringBuffer stringBuffer = new StringBuffer();
-		Matcher matcher = TAG_REGEXP.matcher(str);
-		while (matcher.find())
+		if (currentMessage == null)
 		{
-			matcher.appendReplacement(stringBuffer, "");
-			String match = matcher.group(0);
-			switch (match)
+			return;
+		}
+
+		boolean removed = false;
+		MessageNode messageNode = currentMessage.getMessageNode();
+		final int createdTick = currentMessage.getTick();
+
+		if (client.getTickCount() > createdTick + (config.emojiEmitDuration() * 1000 / GAME_TICK_LENGTH) || force)
+		{
+			ChatLineBuffer ccInfoBuffer = client.getChatLineMap().get(ChatMessageType.GAMEMESSAGE.getType());
+			if (ccInfoBuffer != null)
 			{
-				case "<lt>":
-				case "<gt>":
-					stringBuffer.append(match);
-					break;
+				ccInfoBuffer.removeMessageNode(messageNode);
+				removed = true;
+				currentMessage = null;
 			}
 		}
-		matcher.appendTail(stringBuffer);
-		return stringBuffer.toString();
+
+		// don't force a redraw if we're about to add the message back
+		if (removed && !force)
+		{
+			clientThread.invoke(() -> client.runScript(ScriptID.BUILD_CHATBOX));
+		}
+	}
+
+	private String getEmojiMessage()
+	{
+		ChatMessageBuilder builder = new ChatMessageBuilder();
+		ChatMessageBuilder spacer = new ChatMessageBuilder();
+		Emoji[] allEmoji = Emoji.values();
+
+		FontTypeFace fontFace = client.getWidget(WidgetInfo.CHATBOX_INPUT).getFont();
+		for (int i = 1; i < allEmoji.length + 1; i++)
+		{
+			Emoji emoji = allEmoji[i - 1];
+			builder.img(modIconsStart + emoji.ordinal());
+
+			spacer.append("  " + Text.unescapeTags(emoji.getTrigger()));
+
+			while (fontFace.getTextWidth(spacer.build()) < 55)
+			{
+				spacer.append(" ");
+			}
+
+			if (i % EMOJI_PER_LINE == 0)
+			{
+				spacer.append("\n");
+			}
+
+			builder.append(Text.unescapeTags(spacer.build()));
```

Comment:
I'm not sure what's going on here. You can see I moved the instantiation to inside the loop. When I try to swap to a `StringBuilder` I get a continual loop here.
 ```
while (fontFace.getTextWidth(spacer.build()) < 55)
{
	spacer.append(" ");
}
```

and I can't figure out why. When I compare the value of `fontFace.getTextWidth(spacer.build()` before and after appending some spaces, the width is indeed increasing. Why would this loop forever then?

Otherwise, I've addressed all feedback so far except using `StringBuilder` instead of `ChatMessageBuilder`

Reply:
Yep, not sure what's going on here. When I use a `StringBuilder#append(String)` to add spaces in this loop, the `FontTypeFace fontFace.getTextWidth(spacer.build())` call isn't increasing in value. Even when testing against a string with length > 100, `getTextWidth()` is returning the same value as though the string is getting trimmed or something. Not sure what to do here.

Reply:
You should set a breakpoint here and figure out why it only works before making that change.